File: .20/main.go
package main

import (
	"fmt"
)

func isValid(s string) bool {
	stack := []rune{}

	mapping := map[rune]rune{
		')':'(',
		'}':'{',
		']':'[',
	}

	for _, char := range s {
		if char == '(' || char == '{' || char == '[' {
			stack = append(stack, char)
		} else if (char == ')' || char == '}' || char == ']') {
			if len(stack) == 0 {
				return false
			}

			top := stack[len(stack) - 1]
			stack = stack[:len(stack) - 1]

			if mapping[char] != top {
				return false
			}
		}
	}
	return len(stack) == 0
}


func main() {
	input := "()"
	ans := isValid(input)

	fmt.Println(ans)
}

File: .20/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        # スタック構造の使用
        stack = []

        # ハッシュテーブルを使う
        hash_table = {
            ')': "(",
            "}": "{",
            "]": "[",
        }

        for char in s:
            # charが空きかっこなら
            if char in hash_table.values():
                stack.append(char)

            # charが閉じかっこなら
            if char in hash_table.keys():
                if not stack or hash_table[char] != stack.pop():
                    return False

        return not stack
